lib/seahorn/CexExeGenerator.cc

```diff
@@ -37,122 +37,130 @@ bool extractArrayContents(Expr e, kv &out, Expr &defaultValue) {
   }
   return true;
 }
-} // namespace utils
 
+/* InstVisitor that visits all CallSites in a Bmc Trace;
+  if the CallSite calls an external non-deterministic function that needs to
+  be synthesized, push evaluated function return value or modifed pointer value
+  into Cex generator
+ */
 template <class Trace>
-void CexExeGenerator<Trace>::storeMemHavoc(unsigned loc, const Function *func,
-                                           ImmutableCallSite cs,
-                                           ImmutableCallSite prevCS) {
-  const Value *prevCV = prevCS.getCalledValue();
-  const Function *prevF = dyn_cast<Function>(prevCV->stripPointerCasts());
-  if (!(prevF && prevF->getName().equals("shadow.mem.load"))) {
-    LOG("cex", errs() << "Skipping harness for memhavoc"
-                      << " because shadow.mem.load cannot be found\n");
-    return;
-  }
-  // get memhavoc content from corresponding shadow mem region
-  auto *shadowMemI = dyn_cast<Instruction>(prevCS.getArgOperand(1));
-  if (!shadowMemI)
-    return;
-  Expr shadowMem = m_trace.eval(loc, *shadowMemI, true);
-  // get memhavoc size from second operand of memhavoc
-  const Value *sizeArg = cs.getArgOperand(1);
-  Expr size;
-  if (auto *sizeI = dyn_cast<Instruction>(sizeArg)) {
-    size = m_trace.eval(loc, *sizeI, true);
-  } else if (auto *sizeConst = dyn_cast<ConstantInt>(sizeArg)) {
-    expr::mpz_class sz = toMpz(sizeConst);
-    size = expr::mkTerm<expr::mpz_class>(sz, m_trace.engine().efac());
-  } else {
-    LOG("cex",
-        errs() << "unhandled Value of memhavoc size: " << *sizeArg << "\n");
-    return;
-  }
-  // get info of ptr to havoc
-  const Value *hPtrAlloc = cs.getArgOperand(0)->stripPointerCasts();
-  Expr hPtrStart;
-  if (auto *hPtrAllocInst = dyn_cast<Instruction>(hPtrAlloc)) {
-    hPtrStart = m_trace.eval(loc, *hPtrAllocInst, true);
-  } else {
-    LOG("cex",
-        errs() << "unhandled Value of memhavoc ptr: " << *hPtrAlloc << "\n");
-    return;
+class BmcTraceVisitor : public InstVisitor<BmcTraceVisitor<Trace>> {
+  CexExeGenerator<Trace> &m_cex;
+  unsigned m_loc;
+
+public:
+  BmcTraceVisitor(CexExeGenerator<Trace> &cex) : m_cex(cex) {}
+
+  void setLoc(unsigned loc) { m_loc = loc; }
+
+  void visitMemhavoc(CallSite CS) {
+    Instruction &I = *CS.getInstruction();
+    auto *CF = getCalledFunction(CS);
+    // previous instruction should be
+    // shadow.mem.load(i32 x, i32 %sm_n, i8* null)
+    if (I.getPrevNonDebugInstruction() == nullptr)
+      return;
+    const Instruction *prevI = I.getPrevNonDebugInstruction();
+    if (const CallInst *prevCI = dyn_cast<CallInst>(prevI)) {
+      ImmutableCallSite prevCS(prevCI);
+      const Function *prevF = prevCS.getCalledFunction();
```

Comment:
skipping strip pointer cast should be fine here `shadow.mem.load` is prepended directly